### ISSUE TITLE
ERROR in Illegal state: symbol without members expected when using AOT in 4.3.0

### ISSUE THREAD

**Issue by aleris** [angular/angular#18170]
## I'm submitting a...
<pre><code>
[x] Regression
[ ] Bug report 
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request
</code></pre>

## Current behavior
Getting error when building for production (AOT enabled) with a certain pattern of imports/exports (see below). The error reported is:
```
ERROR in Illegal state: symbol without members expected, but got {"filePath":"snip/test/ng/aottest/src/app/shared/role-co
nstants.ts","name":"RoleConstants","members":["USER"]}.

ERROR in ./src/main.ts
Module not found: Error: Can't resolve './$$_gendir/app/app.module.ngfactory' in 'snip\test\ng\aottest\src'
 @ ./src/main.ts 4:0-74
 @ multi ./src/main.ts
```
Worked ok in 4.2.6, does not work in 4.3.0.

## Expected behavior
ng build --prod should build correctly without rising the error.

## Minimal reproduction of the problem with instructions
download [symbol-without-members-error-test-a.zip](https://github.com/angular/angular/files/1152729/symbol-without-members-error-test-a.zip)
npm install
ng build --prod

The project contain a newly generated project (with ng new) with a number of files added to reproduce the error:
shared/const.ts - exports a constant field
shared/index.ts - exports consts above
intermediate.ts - imports consts exports aggregated data
index.ts - exports intermediate above
module.ts - is calling RouterModule.forRoot([...intermediate] with intermediate imported from above.

If `import { Consts } from './shared';` is changed to `import { Consts } from './shared/consts';` in intermediate.ts the problem does not manifest anymore.

## What is the motivation / use case for changing the behavior?
n/a

## Environment

<pre><code>
Angular version: 4.3.0
(worked ok in 4.2.6.)

For Tooling issues:
- Node version: v7.10.0
- Platform: Windows

</code></pre>

**Comment by manekinekko** [comment]
@aleris I can't reproduce this issue. See screenshot below. What version of the CLI are you using?
 
![image](https://user-images.githubusercontent.com/1699357/28271551-24763672-6b09-11e7-8aaa-60c644f8abc6.png)

**Comment by deepu105** [comment]
We were using 1.2.1 in JHipster

**Comment by deepu105** [comment]
I made few tests with this sample
With Global CLI 1.2.1 and local 1.0.0
![image](https://user-images.githubusercontent.com/1107223/28273947-df755542-6b0f-11e7-891d-e14519fd92e4.png)
With global and local CLI 1.2.1
![image](https://user-images.githubusercontent.com/1107223/28274024-09aeee68-6b10-11e7-9af5-3010316aa6fc.png)
With local CLI 1.3.0-beta.1
![image](https://user-images.githubusercontent.com/1107223/28274117-420eb644-6b10-11e7-9052-fb6828e69ca4.png)
With global and local CLI 1.3.0-beta.1
![image](https://user-images.githubusercontent.com/1107223/28274192-82a3b16e-6b10-11e7-9d9c-9a1e94346841.png)

Could it be possible that you have a different angular version linked or something?

**Comment by aleris** [comment]
```
ng -v
@angular/cli: 1.0.0
node: 7.10.0
os: win32 x64
@angular/common: 4.3.0
@angular/compiler: 4.3.0
@angular/core: 4.3.0
@angular/forms: 4.3.0
@angular/http: 4.3.0
@angular/platform-browser: 4.3.0
@angular/platform-browser-dynamic: 4.3.0
@angular/router: 4.3.0
@angular/cli: 1.0.0
@angular/compiler-cli: 4.3.0
```
After:
`npm install @angular/cli@1.3.0-beta.1`

![a](https://user-images.githubusercontent.com/1011245/28274486-b990da40-6b19-11e7-9ba8-59820149891f.png)

Same with 1.2.1 (@latest)

**Comment by cjamaldine** [comment]
hello,

it works with :
@angular/compiler-cli: 4.2.3

**Comment by tmakin** [comment]
I'm also getting this on an enum export when using 4.3.0:
```
ERROR in Illegal state: symbol without members expected, but got {"filePath":"snip/src/app/core/models/index.ts","name":"Role","members":["Advisor"]}.
```

Downgrade to 4.2.6 fixes it

**Comment by janousek** [comment]
The error rises because of the commit https://github.com/angular/angular/commit/ddb766e456f4d629ffe344d1178476c8c8657375

Concretely because of the new condition `references > 0` in this if:

                    // Stop simplification at builtin symbols or if we are in a reference context
                    if (expression === self.injectionToken || expression === self.opaqueToken ||
                        self.conversionMap.has(expression) || references > 0) {
                        return expression;
                    }

**Comment by rimlin** [comment]
+1 this error with enums like:
```
export enum MyEnum {
  val1 = <any>"val1",
  val2 = <any>"val2"
}
```

**Comment by cdllqos** [comment]
Same error as this, I just new a project and exec 'ng build --prod', and then the error comes :). By the way, I install npm package by cnpm
![image](https://user-images.githubusercontent.com/25027791/28528744-2f036e62-70c1-11e7-809c-11ed5ee79e80.png)

![image](https://user-images.githubusercontent.com/25027791/28528732-23db3d26-70c1-11e7-84b4-8da360f7d95d.png)

**Comment by montella1507** [comment]
@chuckjaz can you give us some advice, please?

**Comment by florinr1** [comment]
I have the same error after upgrading angular from 4.1.3 to 4.3.1 and CLI from 1.1.0 to 1.2.3.
`export enum ModDeschidere {
  CREATE = 0,
  VIEW = 1,
  EDIT = 2
}`

> ERROR in Illegal state: symbol without members expected, but got {"filePath":"src/app/core/modDeschidere.ts","name":"ModDeschidere","members":["CREATE"]}.

I don't know what to do / test.

**Comment by janousek** [comment]
For everyone who needs temporary fix for this error. Just manually modify this file:

> node_modules\@angular\compiler\bundles\compiler.umd.js

Find this part of code:

                    // Stop simplification at builtin symbols or if we are in a reference context
                    if (expression === self.injectionToken || expression === self.opaqueToken ||
                        self.conversionMap.has(expression) || references > 0) {
                        return expression;
                    }

And modifi it to look like this:

                    // Stop simplification at builtin symbols or if we are in a reference context
                    if (expression === self.injectionToken || expression === self.opaqueToken ||
                        self.conversionMap.has(expression)) {
                        return expression;
                    }

Everything will work.

**Comment by montella1507** [comment]
@janousek  unfortunately it is not possible (it is dangerous) in CI environment when you want to build "clean build". So i would suggest temporary downgrade to older angular as temporary solution.

**Comment by erlloyd** [comment]
I still get the error even after downgrading my @angular packages to 4.2.6

<img width="1059" alt="screenshot 2017-07-29 22 29 04" src="https://user-images.githubusercontent.com/2145299/28750335-7392bcbe-74ad-11e7-98ef-cd628d70dd08.png">

And @janousek I confirmed the code you pointed out is rolled back. Does anyone have any updates / ideas on this?

**Comment by jesussobrino** [comment]
I tried Angular **4.3.2** and this bug it's still been reproduced.

**Building with Angular 4.3.2 AoT:**
> Error: Illegal state: symbol without members expected, but got {"filePath":"/my-model.model.ts","name":"ModelCode","members":["ModelData"]}.

**Building with Angular 4.3.2 without AoT:**
> [at-loader] Ok, 6.462 sec.

**After downgrading to Angular **4.2.6** with or without AoT:**
[at-loader] Ok, 10.517 sec.

I don't understand why the angular team doesn't care about this, because it's impossible to use the latest AoT Angular version in any CI environment.

**Comment by michal-filip** [comment]
Would anyone have a 4.3.x fork with the workaround in compiler.umd.js so that we don't have to downgrade? But I'm still hoping for a fix or at least some suggested fix instruction from @chuckjaz .

**Comment by OysteinAmundsen** [comment]
Still an issue with 4.3.3.

**Comment by rafaelss95** [comment]
I'm facing the same problem having an enum like this:

```
export enum Role {
  User = 'user',
  Admin = 'admin'
}

```

> ERROR in Illegal state: symbol without members expected, but got {"filePath":"/filePath/enums/role.enum.ts","name":"Role","members":["User"]}.

Using Angular 4.3.2 and CLI 1.2.6:

**Comment by Vladimex** [comment]
I have the same problem with 4.3.3.
When it will be fixed?

**Comment by Mindsers** [comment]
+1 @rafaelss95 

We have the same problem and reverting is not an option: we can't modify the application's code base.

**Comment by michal-filip** [comment]
The simplest way for us was to fix the compiler tarball (based on fix suggested by @janousek) and put it next to our project, then fix references to all angular modules to 4.3.3 and link `@angular/compiler` to the local tarball.
Package.json then looks like this:
```
"dependencies": {
        "@angular/common": "4.3.3",
        "@angular/compiler": "../dependency-overrides/compiler-4.3.3-fix.tgz",
        "@angular/compiler-cli": "4.3.3",
```
zipped tarball:
[compiler-4.3.3-fix.tgz.zip](https://github.com/angular/angular/files/1212861/compiler-4.3.3-fix.tgz.zip)
After this, we no longer get the issue during AOT builds.

Hopefully someone more familiar with the compiler will be able to create a proper fix PR soon, it must be blocking lots of teams from upgrading.

**Comment by OysteinAmundsen** [comment]
Works for me in 4.3.4! 🥇

**Comment by montella1507** [comment]
Almost a month with this very important bug without author reaction :-(

**Comment by yoricktran** [comment]
Running without the --prod flag at the moment, but could you please fix this until I go into production? :3

**Comment by aldo-roman** [comment]
Running with --prod but without --aot worked for me:

`ng build --prod --aot=false`

**Comment by Mindsers** [comment]
@aldo-roman `—aot` functionality is precisely what we need here...

**Comment by aldo-roman** [comment]
@Mindsers sorry, I meant, "I can build by disabling just _AOT_, and not all _prod_ flag"

I do need AOT, too.

**Comment by chuckjaz** [comment]
If the enum reference is inside a provider `useValue` or route's `data` properties, you can work around this by introducing an intermediate exported variable that the template compiler can use instead of inlining the expression.

**Comment by coryrylan** [comment]
Running into this bug as well. This issue looks like it might be related https://github.com/angular/angular/issues/16321

**Comment by DincaRoberto** [comment]
@chuckjaz Is it possible to get an example of the work around you mentioned in your comment?

**Comment by michal-filip** [comment]
@DincaRoberto In my case the following kind of change helped (thanks @chuckjaz):
```
+ export const UiFlag_CONTROL_BOARD_VISIBLE = UiFlag.CONTROL_BOARD_VISIBLE;
const routes: Routes = [
    // ...
   {
         path: 'control-board',
         canActivate: [AuthGuard, FlagGuard],
         data: {
-            flag: UiFlag.CONTROL_BOARD_VISIBLE
+            flag: UiFlag_CONTROL_BOARD_VISIBLE
         }
         // ...
     }
];
```

**Comment by chuckjaz** [comment]
@DincaRoberto Something like example provided @michal-filip is what I was referring to.

**Comment by Vladimex** [comment]
Problem still exists on 4.3.5 when I try to execute: npm run ionic:build --aot
```
ionic-app-script task: "build"
Error: Illegal state: symbol without members expected, but got {"filePath":".../src/components/name-item/name-item.ts","name":"NameItem","members":["loader"]}.
```
Any ideas how to fix or when it will be fixed?
Without AOT ionic app cannot be published...

**Comment by chuckjaz** [comment]
The work-around described above should work for this.

**Comment by piernik** [comment]
@chuckjaz michal-filip's didn't work for me.

**Comment by trotyl** [comment]
Should it be fixed by https://github.com/angular/angular/pull/18905?

**Comment by chuckjaz** [comment]
@trotyl Yes. In 5 the compiler will "lower"  these expressions into exported variables automatically avoiding this issue entirely.

@piernik What exactly did you try?

**Comment by piernik** [comment]
In my `config.module` I import my own `SchematerModule` which is responsible for reactive forms.
In it I define validators and assing static methods:

```
SchematerFieldsModule.forRoot({
	//config
	validators:
		{
			unikalnyTytulKrainy: {
				method: CustomValidators.unikalnyTytulKrainy,
				async: true,
				message: 'Taki tytuł już istnieje'
			},
			unikalnyTytulAtrakcje: {
				method: CustomValidators.unikalnyTytulAtrakcji,
				async: true,
				message: 'Taki tytuł już istnieje'
			},
		}
}),
```
That was my inital code. 
I changed it to: 
```
export const CustomValidators_unikalnyTytulKrainy = CustomValidators.unikalnyTytulKrainy;
export const CustomValidators_unikalnyTytulAtrakcji = CustomValidators.unikalnyTytulAtrakcji;

...
method: CustomValidators_unikalnyTytulKrainy,
method: CustomValidators_unikalnyTytulAtrakcji,
```
In both cases the same error: `ERROR in Illegal state: symbol without members expected`


Here are my static methods
```
export class CustomValidators {
    static unikalnyTytulAtrakcji(fieldControl: FormControl): Observable<any> | Promise<any> {
        const apiService = (this as any).injector.get(ApiService);
        return CustomValidators.unikalnyTytul(fieldControl, 'atrakcje', apiService, this);
    }

    static unikalnyTytulKrainy(fieldControl: FormControl): Observable<any> | Promise<any> {
        const apiService = (this as any).injector.get(ApiService);
        return CustomValidators.unikalnyTytul(fieldControl, 'krainy', apiService, this);
    }

    static unikalnyTytul(fieldControl: FormControl, obiekt: string, apiService: ApiService, self: any): Observable<any> | Promise<any> {
        let id;
        if (fieldControl.root && fieldControl.root.value) {
            const schematerConfig: SchematerObjectsConfigService = self.injector.get(SchematerObjectsConfigService);
            id = fieldControl.root.value[schematerConfig.getObject(obiekt).primaryField];
        }
        return apiService.get('walidacja/tytul-' + obiekt, {tytul: fieldControl.value, bezId: id})
            .map(resp => {
                if (resp.valid) {
                    return null;
                } else {
                    return {unikalnyTytulAtrakcje: true};
                }
            });
    }
}
```

**Comment by chuckjaz** [comment]
The work around won't work in this case.

Currently we only treat the fields `useValue`, `useFactory`, and `data` as special. We are planning on adding `useClass` and `animations`.

Starting in 5.0.0-beta.6 the compiler will automatically introduce exported variables (in the generated code, not the .d.ts file) to allow the factory to have access to these values.

Given that none of this will help your example, we are going to 1) fix references to enums (this issue) for 5.0.0 final, and 2) investigate if the rewriting should be more general because it would still fail if you used an unrecognized call in this code which also should be allowed.

**Comment by piernik** [comment]
@chuckjaz Thanks for response.

Can You give me a working solution? Example, article or something?

**Comment by nerumo** [comment]
tl;dr; I resolved my issue by directly re export the file in the root index.ts instead of the index.ts from the subdir where the file lives.

**Comment by OysteinAmundsen** [comment]
In my case this error happens on a plain enum. Not sure if the workaround is applicable for my case, or how I would go about implementing this workaround in my particular case...

```
export enum Role {
  Admin = 99, Organizer = 90, Secretariat = 80, Club = 50, User = 10
}

export const RoleNames: [{id: number, name: string}] = [
  {id: Role.Admin,       name: 'Admin'},
  {id: Role.Organizer,   name: 'Organizer'},
  {id: Role.Secretariat, name: 'Secretariat'},
  {id: Role.Club,        name: 'Club'},
  {id: Role.User,        name: 'User'},
]

export interface IUser extends IBelongsToClub  {
  id: number;
  name: string;
  email: string;
  password: string;
  role?: Role;
}
```

`ERROR in Error: Illegal state: symbol without members expected, but got {"filePath":"./client/app/model/IUser.ts","name":"Role","members":["Organizer"]}.`

**Comment by jdrake3** [comment]
Is this the same as [6443](https://github.com/angular/angular-cli/issues/6443) ?

**Comment by Sliverb** [comment]
For those that want a workaround till angular 5 rolls around (or this gets fixed). I created a repo with @janousek fix here - https://github.com/Sliverb/angular-compiler.git

I have automated builds set up and this workaround allows things to stay automated till a fix is out

Usage:
In package.json use the src above for the compiler.
"@angular/compiler": "git+https://github.com/Sliverb/angular-compiler.git",

I'm still trying to understand the reason behind the change and what it does, so if anyone knows. Pls share so we can update our code accordingly.

Thanks guys for leading to this workaround

**Comment by junglebarry** [comment]
Hi,

I’m vaguely aware that Angular 4 is going to get LTS maintenance. However, the answer for this issue seems to be “wait for v5”.

Are there plans to fix this issue for v4, because if not, many people using standard typescript language features seem like they’ll be unable to access the LTS patches past 4.2.6.

Apologies if I’ve completely misunderstood the advice and workarounds.

Thanks in advance!

**Comment by NateWarner** [comment]
Just updated to release 4.4.4, which according to the [changelog](https://github.com/angular/angular/commit/bed8ac70f73f47fd03333c9ee47a0439254e3aff) included the fix for this, and the issue is still present when using an `enum` on a route.

**Comment by Vladimex** [comment]
Updated to Angular 4.4.4
Error is still there: "Error: Illegal state: symbol without members expected, but got..."

**Comment by OysteinAmundsen** [comment]
4.4.4 Works on my environment (windows and ubuntu in docker), enums and all! 

Me === Happy!

**Comment by GeorgeKnap** [comment]
just getting this error.
```
ERROR in Error: Illegal state: symbol without members expected, but got {"filePath":"C:/Development/Intens.eCall/trunk/Intens.eCall.Website/src/app/shared/models/roles.enum.ts","name":"Role","members":["eCallUserAdministrator"]}.
    at StaticSymbol.assertNoMembers (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:794:19)
    at AotSummaryResolver.resolveSummary (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:18948:22)
    at CompileMetadataResolver._loadSummary (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:8822:49)
    at CompileMetadataResolver._getEntryComponentMetadata (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9529:31)
    at C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9516:31
    at Array.forEach (native)
    at CompileMetadataResolver._getEntryComponentsFromProvider (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9515:30)
    at C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9488:83
    at Array.forEach (native)
    at CompileMetadataResolver._getProvidersMetadata (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9451:19)
    at C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9453:23
    at Array.forEach (native)
    at CompileMetadataResolver._getProvidersMetadata (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9451:19)
    at C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9099:63
    at Array.forEach (native)
    at CompileMetadataResolver.getNgModuleMetadata (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9090:49)
    at CompileMetadataResolver.getNgModuleSummary (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9035:35)
    at C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9105:55
    at Array.forEach (native)
    at CompileMetadataResolver.getNgModuleMetadata (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9090:49)
    at CompileMetadataResolver.getNgModuleSummary (C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9035:35)
    at C:\Development\Intens.eCall\trunk\Intens.eCall.Website\node_modules\@angular\compiler\bundles\compiler.umd.js:9105:55
npm ERR! Windows_NT 10.0.15063
```
at enum: 
```
export enum Role {
    eCallUserAdministrator,
    MSD,
    IVS,
    PSAP
}
```


packages:
```
  "dependencies": {
    "@angular/core": "5.0.0-beta.7",
    "@angular/common": "5.0.0-beta.7",
    "@angular/animations": "5.0.0-beta.7",
    "@angular/platform-browser": "5.0.0-beta.7",
    "@angular/platform-browser-dynamic": "5.0.0-beta.7",
    "@angular/compiler": "5.0.0-beta.7",
    "@angular/flex-layout": "git+https://github.com/angular/flex-layout-builds.git",
    "@angular/forms": "5.0.0-beta.7",
    "@angular/http": "5.0.0-beta.7",
    "@angular/router": "5.0.0-beta.7",
    "@angular/material": "2.0.0-beta.11",
    "@angular/cdk": "2.0.0-beta.11",
    "core-js": "2.4.1",
    "jquery": "3.2.1",
    "rxjs": "5.4.0",
    "zone.js": "0.8.17",
    "servicestack-client": "0.0.39",
    "event-source-polyfill": "0.0.9"
  },
  "devDependencies": {
    "@types/node": "*",
    "@types/lodash": "*",
    "@types/service_worker_api": "*",
    "@types/arcgis-js-api": "4.4.0",
    "@angular/cli": "1.4.4",
    "@angularclass/hmr": "2.1.3",
    "@angular/compiler-cli": "5.0.0-beta.7",
    "@angular/service-worker": "1.0.0-beta.16",
    "codelyzer": "3.2.0",
    "node-sass": "4.5.3",
    "ts-node": "3.3.0",
    "tslint": "5.7.0",
    "typescript": "2.5.3",
    "webpack-bundle-analyzer": "2.9.0"
  }
```

is this a problem related to cli or angular itself?

**Comment by chuckjaz** [comment]
@NateWarner It turns out my fix was incomplete. I fixed some, but not all, of the cases.

If you replace line ~25021 in compiler.umd.js that looks like:
```
var /** @type {?} */ selectTarget = simplify(expression['expression']);
```

with:

```
var /** @type {?} */ selectTarget = simplifyInContext(context, expression['expression'], depth, 0);
```

It should fix the case you are running into.

**Comment by chuckjaz** [comment]
@NateWarner Can you verify that the fix above resolves your issue. If it does I will try to get this into 4.4.5.

**Comment by NateWarner** [comment]
@chuckjaz I can confirm that making this change does fix `enums` on routes. Thanks for the quick response.

**Comment by sunb0002** [comment]
Sorry I'm still getting the issue with latest versions.
It's some enum generated by swagger.

Angular CLI: 1.5.5
Node: 6.10.3
OS: win32 x64
Angular: 4.4.6

 @ ./src/main.ts 3:0-74
 @ multi ./src/main.ts
ERROR in Error: Illegal state: symbol without members expected, but got {"filePath":"C:/Users/xxx/Documents/workspaceWORK/xxx/src/app/shared/backend-api/emart/model/GetUserPermissionsResponseResult.ts","name":"GetUserPermissionsResponseResult","members":["PermissionsEnum","CREATEUPDATEROLE"]}.

**Comment by chuckjaz** [comment]
@sunb0002 Can you create a small repository that reproduces this problem I can clone?

**Comment by sunb0002** [comment]
@chuckjaz Thank you for the prompt response.
I've created it here: https://github.com/sunb0002/angularissue
'ng build' is ok, while 'ng build --aot' gets errors. 

Also noticed that if I aot compile the same code with Angular 5.0.0, I won't get the error.
But anyway we are not rushing on this issue as it's an internal application.
We can persuade client to go without aot.

**Comment by chuckjaz** [comment]
@sunb0002 A work around for your case is to introduce a intermediate exported symbol:

```ts
export const HomePermissions = [PermissionsEnum.CREATEVENDORUSER];

const routes: Routes = [
  {
    path: '**', redirectTo: 'dashboard',
    data: { title: 'Home', accessPermissions: HomePermissions }
  }
];
```

This type of rewriting is performed automatically in 5.0 and is transformed into something like,

```ts
export const ɵ1 = { title: 'Home', accessPermissions: [PermissionsEnum.CREATEVENDORUSER] };

const routes: Routes = [
  {
    path: '**', redirectTo: 'dashboard',
    data: ɵ1
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes, { useHash: true })],
  exports: [RouterModule]
})
export class RoutesRoutingModule { }
```

I will rework the assert into a better error message for 4.x.x. We will not support indexing into members of an namespace but, at least, we should give a reasonable error message instead of an internal assert violation.

**Comment by sunb0002** [comment]
@chuckjaz Thanks a lot for the help. It's working now.
